File: spring/Rope_t.cs
using System;

namespace spring
{
    public class Rope_t
    {
        public Node_t[] Nodes;
        public Element_t[] Elements;
        public float[][][] load;
        public Rope_t(props Props, float[][][] _load)
        {
            Nodes = new Node_t[Props.nodes];
            load = _load;
            SetupNodesPositions(Props.Counts, Props.initDrop, Props.L, Props.E, Props.D);
            EvalElements(Props.D, Props.ro);
        }
        private void SetupNodesPositions(int Counts, float initDrop, float L, float E, float D)
        {
            float dl = L / Nodes.Length;
            Nodes[0] = new Node_t(Counts, new float[3] { 0, initDrop * (float)Math.Pow((0 * dl) - (L - dl) / 2, 2) + 1E-3f, 0 }, NodeFreedom.xyz, NodeLoad.f, 0, new int[1] { 1 }, E, D);
            for (int i = 1; i < Nodes.Length - 1; i++)
            {
                Nodes[i] = new Node_t(Counts, new float[3] { i * dl, initDrop * (float)Math.Pow((i * dl) - (L - dl) / 2, 2) + 1E-3f, 0 }, NodeFreedom.xyz, NodeLoad.none, i, new int[2] { i - 1, i + 1 }, E, D);
            }
            Nodes[Nodes.Length - 1] = new Node_t(Counts, new float[3] { (Nodes.Length - 1) * dl, initDrop * (float)Math.Pow(((Nodes.Length - 1) * dl) - (L - dl) / 2, 2) + 1E-3f, 0 }, NodeFreedom.xyz, NodeLoad.f, Nodes.Length - 1, new int[1] { Nodes.Length - 2 }, E, D);
        }
        private void EvalElements(float _D, float ro)
        {
            foreach (var element in Elements)
            {
                element.A = (float)Math.PI * (float)Math.Pow(_D, 2) / 4;
                element.I = maf.P3(element.A) / 12f;
                float L = crds.GetTotL(Nodes[element.n1].tm[0][(int)N.p], Nodes[element.n1].tm[0][(int)N.p]);
                if (L == 0)
                {
                    throw new Exception();
                }
                element.m = ro * element.A * L / 2;
            }
        }

        public void IterateOverNodes(int t, float dt)
        {
            foreach (var node in Nodes)
            {
                /*get loading*/
                vectr.Plus(ref node.tm[t][(int)node.LoadType], load[node.NodeID][t]);
                /*calc force*/
                node.GetForces(Nodes, t);
                /*integrate*/
                node.Integrate(t, t - 1, dt);
            }
        }
    }
}

File: mechLIB_CPP/Rope_t.cpp
#include "pch.h"
#include "Rope_t.h"
#include <algorithm>    // std::for_each
#include <execution>
#include "maf.hpp"

void Rope_t::init(mechLIB_CPP::ModelPropertiesNative* props)
{

	L = std::vector<float>(props->Counts);
	L[0] = props->L;
	NodesSize = props->nodes;
	Nodes = std::vector<Node_t>(NodesSize);
	ElementsSize = props->nodes - 1;
	Elements = std::vector<Element_t>(ElementsSize);
}

void Rope_t::SetupNodesPositions(mechLIB_CPP::ModelPropertiesNative* props)
{
	SetupNodesPositions(props, DirectX::SimpleMath::Vector3{ 0,0,0 }, DirectX::SimpleMath::Vector3{ props->L,0,0 });
}

void Rope_t::SetupNodesPositions(mechLIB_CPP::ModelPropertiesNative* props, DirectX::SimpleMath::Vector3 startCoord,
	DirectX::SimpleMath::Vector3 endCoord)
{
	size_t lastNode = NodesSize - 1;
	float dl = props->L / lastNode;
	float cosA = endCoord.x / props->L;
	float sinA = endCoord.y / props->L;
	Nodes[0].init(props->Counts, startCoord,
		mechLIB_CPP::NodeFreedom::xyz, mechLIB_CPP::NodeLoad::none, std::vector<Element_t*>{ &Elements[0] });
	for (size_t i = 1; i < lastNode; i++)
	{
		DirectX::SimpleMath::Vector3 flatC(i * dl, 0, 0);
		DirectX::SimpleMath::Vector3 coords(

			(flatC.x - startCoord.x) * cosA - (flatC.y - startCoord.y) * sinA + startCoord.x,
			(flatC.x - startCoord.x) * sinA + (flatC.y - startCoord.y) * cosA + startCoord.y, 0
		);
		Nodes[i].init(props->Counts, coords,
			mechLIB_CPP::NodeFreedom::xyz, mechLIB_CPP::NodeLoad::none, std::vector<Element_t*>{ &Elements[i - 1], & Elements[i] });
	}
	Nodes[lastNode].init(props->Counts, endCoord,
		mechLIB_CPP::NodeFreedom::xyz, mechLIB_CPP::NodeLoad::none, std::vector<Element_t*>{ &Elements[lastNode - 1] });
}

void Rope_t::EvalElements(mechLIB_CPP::ModelPropertiesNative* props)
{
	for (size_t i = 0; i < ElementsSize; i++)
	{
		Elements[i].init(&Nodes[i], &Nodes[i + 1], props);
		Elements[i].L[0] = Elements[i].GetOwnLength(0);
	}
}

void Rope_t::StepOverNodes(size_t t, float Re, float dt)
{
//#pragma omp parallel for
	for (size_t i = 0; i < NodesSize; i++)
	{
		float m = 0;
		float c = 0;
		Nodes[i].GetPhysicParam(t - 1, Re, m, c);
		Nodes[i].GetForces(t, m, c);
		Nodes[i].CalcAccel(t, m);
		/*integrate*/
		Nodes[i].Integrate(t, t - 1, dt);
	}
}
